Add plugin install and uninstall commands

carolynvs opened the pull request:
# What does this change
Support installing and uninstall plugins with `porter plugins install` and `porter plugins uninstall`.

The commands are exactly the same as for mixins only they install plugins. 😅 To assist with this I have moved the majority of the logic that was originally for mixins into a new package `pkg/pkgmgmt` "package management" that handles downloading, installing, listing, updating, etc packages (which could be a mixin or a plugin).

# What issue does it fix
#771

# Notes for the reviewer
* This adds plumbing for a lot of the other plugin commands but for now I just care about install for our demo to the CNAB group on Wednesday.
* This impacts PR https://github.com/deislabs/porter/pull/868
* I am following up with documentation explaining plugins in general and plugins vs mixins.

# Checklist
- [x] Unit Tests
- [x] Documentation
  - [ ] Documentation Not Impacted

dev-drprasad commented on docs/content/cli/plugins_uninstall.md:
````diff
@@ -0,0 +1,39 @@
+### Examples
+
+```
+  porter mixin uninstall azure
````
Problem with copy/paste :)

carolynvs replied:
oops! Thank you for catching, I will review all of the documentation I copy/pasted for that error.

vdice commented on pkg/mixin/helpers.go:
```diff
@@ -5,42 +5,49 @@ import (
 
-func (p *TestMixinProvider) List() ([]Metadata, error) {
-	mixins := []Metadata{
+func (p *TestMixinProvider) List() ([]pkgmgmt.Metadata, error) {
```
Do we want to change the name from `TestMixinProvider` to just, say, `TestMixin`?  Noticed the 'real' former `MixinProvider` looks to be changed to just `Mixin`...

carolynvs replied:
I believe this name is more consistent with our other test providers naming `Test+Struct|InterfaceName`, and follows what @dev-drprasad did in their PR:

Here's some of the other test stubs we have for reference
https://github.com/deislabs/porter/blob/b27f4a04f9a143c0c853ea37ba71b07d597bad6b/pkg/porter/helpers.go#L44-L55

vdice commented on pkg/pkgmgmt/client/filesystem.go:
```diff
@@ -111,33 +129,61 @@ func (fs *FileSystem) GetVersionMetadata(m mixin.Metadata) (*mixin.VersionInfo,
+		return "", errors.Wrapf(err, "%s %s not accessible at %s", fs.PkgType, name, pkgDir)
+	}
+	if !dirExists {
+		return "", fmt.Errorf("mixin %s not installed", name)
```
stray `mixin` here

vdice commented on pkg/pkgmgmt/client/runner.go:
```diff
@@ -7,25 +7,25 @@ import (
-		runtime:  runtime,
-		mixinDir: mixinDir,
+		Context: context.New(),
+		mixin:   mixin,
```
should this `mixin` field name be updated?  `pkg` or `package`?

vdice commented on pkg/pkgmgmt/client/runner.go:
```diff
@@ -34,7 +34,7 @@ func (r *Runner) Validate() error {
 		return errors.New("mixin not specified")
```
Assuming the field name is updated, etc., we should update the error text to use the generic `package` as opposed to `mixin`...

carolynvs replied:
You're right, I'll do a passthrough on the entire pkgmgmt package for "mixin" and fix the text.

dev-drprasad commented on pkg/pkgmgmt/client/delete.go:
```diff
@@ -0,0 +1,44 @@
+	}
+
+	if fs.Debug {
+		fmt.Fprintf(fs.Out, "Unable to find requested %s\n", fs.PkgType)
```
I don't know whether its big deal or not but we used to prefix debug logs with `DEBUG`

carolynvs replied:
I am breaking my own rules, thank you for calling that out. 🙇‍♀

carolynvs replied:
After looking at this one further, I think it's fine to not have the DEBUG: prefix. In the other places it was just to help call out that we were dumping variables.

Here it's not needed as we are just printing additional English text when in debug mode.

However I wasn't following the `Err` convention, and the message was lacking, so I improved it.

carolynvs commented on pkg/pkgmgmt/pkgmgmt.go:
```diff
@@ -0,0 +1,51 @@
+
+	// GetVersion is the obsolete form of retrieving mixin version, e.g. exec version, which returned an unstructured
+	// version string. It will be deprecated soon and is replaced by GetVersionMetadata.
+	GetVersion(Metadata) (string, error)
```
Yes I believe we can get rid of it now 👍, all of the mixins support the new signature now. Let's do that in a separate PR from this large refactoring so that it's easier to track down in history later.

carolynvs commented on pkg/pkgmgmt/pkgmgmt.go:
```diff
@@ -0,0 +1,51 @@
+	File string
+
+	// PreRun allows the caller to tweak the command before it is executed.
+	// This is only necessary if being called directory from a runner, if
```
Oops, I intended to say "directly"

carolynvs commented:
> Do we have implementations associated to mixins also ? If not, how are we going to use pkgmgmt.Metadata for both mixins and plugins?

This is why the PR is still in draft, it's not complete yet. I should have called that out in the review notes. I don't have a solution yet just an idea in my head. 😀 

Let me try your suggestion for changing the PackageManager interface. It looks good, should work better than what we have now.

carolynvs commented:
I've updated the PR with @dev-drprasad 's feedback on the `PackageManager` interface and addressed how to share it between mixins and plugins (when plugins have a different set of data returned by version).

I haven't finished the tests yet and it needs the cmd doc updated still.

dev-drprasad commented:
@carolynvs Are you waiting for second round of review ?

carolynvs commented:
@dev-drprasad If you would like to provide feedback on how the new commits looks with your suggestions, there's still time. I will be working on adding any missing tests and documentation today. 

The PR isn't being held up waiting for your review though. It will need approval @vdice in order to merge.

carolynvs commented:
As part of reconciling `mixin list` and `plugin list`, I implemented `plugin show` and moved displaying details about plugin implementations into show. I think that fits better with our patterns elsewhere in porter (my mistake for asking for implementations in list to begin with).

```console
$ porter plugin show azure
Name: azure
Version: v0.1.1-9-gf54d8a2
Commit: f54d8a2
Author: Porter Authors

---------------------------
  Type     Implementation
---------------------------
  storage  blob
  secrets  keyvault
```

dev-drprasad commented:
> As part of reconciling `mixin list` and `plugin list`, I implemented `plugin show` and moved displaying details about plugin implementations into show. I think that fits better with our patterns elsewhere in porter (my mistake for asking for implementations in list to begin with).
> 
> ```
> $ porter plugin show azure
> Name: azure
> Version: v0.1.1-9-gf54d8a2
> Commit: f54d8a2
> Author: Porter Authors
> 
> ---------------------------
>   Type     Implementation
> ---------------------------
>   storage  blob
>   secrets  keyvault
> ```

Make sense 👍

carolynvs commented:
/azp run porter

carolynvs commented:
/azp run porter
/azp run porter-integration

carolynvs commented:
/azp run

vdice commented on pkg/pkgmgmt/client/filesystem.go:
```diff
@@ -0,0 +1,144 @@
+	*config.Config
+
+	// PackageType is the type of package managed by this instance of the
+	// package manager It must also correspond to the directory name in
```
Missing a period to end `package manager`

vdice commented on pkg/pkgmgmt/install_test.go:
```diff
@@ -0,0 +1,115 @@
+	t.Run("version specified", func(t *testing.T) {
+		opts := InstallOptions{Version: "canary"}
+		opts.defaultVersion()
+		assert.Equal(t, "canary", opts.Version, "defaultVersion should not overwrite the user's choid")
```
`s/choid/choice`?

carolynvs replied:
I am pro choid

carolynvs commented on docs/config.toml:
```diff
@@ -140,6 +140,18 @@ defaultContentLanguage = "en"
     identifier = "cli-archive"
     weight = 161
     parent = "cli"
+  [[menu.main]]
```
oops! yes thanks 🙇‍♀
